File: play/shared/abilityCast.ts
/**
 * Ability tier casting helpers for play (and shared rules with the battle sim).
 * Alias map is kept in commanderEffectAliases.json (also loaded by battleSim.mjs).
 */
import aliasesJson from './commanderEffectAliases.json' with { type: 'json' }

export type AbilityDef = {
  name: string
  type: string | null
  cost: string | null
  costAmount: number | null
  costResource: string | null
  description: string | null
  usedBy: string | null
  cooldown: number | null
}

export const COMMANDER_EFFECT_ALIASES: Record<string, string> = {
  ...(aliasesJson as Record<string, string>),
}

export function resolveEffectAbilityName(abilityName: string): string {
  return COMMANDER_EFFECT_ALIASES[abilityName] || abilityName
}

export function isUltimateAbility(def: AbilityDef | null | undefined): boolean {
  if (!def) return false
  const kind = (def.type || '').trim()
  const cost = (def.cost || '').trim().toLowerCase()
  return kind === 'Ultimate' || cost === 'ultimate'
}

export function isPassiveAbility(def: AbilityDef | null | undefined): boolean {
  if (!def) return false
  const kind = (def.type || '').trim()
  const cost = (def.cost || '').trim().toLowerCase()
  return kind === 'Passive' || cost === 'passive'
}

export function abilityCostsCc(def: AbilityDef | null | undefined): boolean {
  if (!def) return false
  if ((def.costResource || '').trim().toUpperCase() === 'CC') return true
  const upper = (def.cost || '').toUpperCase()
  if (!upper.includes('CC') || upper.includes('COMPANY')) return false
  return /\b\d+\s*CC\b/.test(upper) || upper.trim() === 'CC'
}

/** Whether a unit kind may cast this ability under the Used By taxonomy. */
export function casterMayUseAbility(
  def: AbilityDef,
  casterKind: 'commander' | 'officer' | 'unit',
): boolean {
  if (isPassiveAbility(def)) return false
  const usedBy = (def.usedBy || '').trim()
  if (casterKind === 'commander') {
    if (isUltimateAbility(def)) return true
    return usedBy === 'Commander'
  }
  if (casterKind === 'officer') {
    if (isUltimateAbility(def) || abilityCostsCc(def)) return false
    return usedBy === 'Officer' || usedBy === 'Both'
  }
  if (isUltimateAbility(def) || abilityCostsCc(def)) return false
  return !usedBy || usedBy === 'Unit' || usedBy === 'Both'
}

export type AbilitySpend =
  | { pool: 'commanderAp'; amount: number }
  | { pool: 'commanderCc'; amount: number }
  | { pool: 'companyAp'; amount: number }
  | { pool: 'none'; amount: 0 }

/** Resolve which pool an active ability spends (by caster role + cost). */
export function abilitySpendForCaster(
  def: AbilityDef,
  casterKind: 'commander' | 'officer' | 'unit',
): AbilitySpend | { error: string } {
  if (isPassiveAbility(def)) return { error: 'Passives are always on — do not cast them.' }
  if (isUltimateAbility(def)) return { pool: 'none', amount: 0 }

  const amount = Math.max(0, Number(def.costAmount) || 0)
  if (abilityCostsCc(def)) {
    if (casterKind !== 'commander') {
      return { error: 'CC abilities are commander-only.' }
    }
    return { pool: 'commanderCc', amount: Math.max(1, amount || 1) }
  }

  const resource = (def.costResource || '').trim().toUpperCase()
  if (resource === 'AP' || /\bAP\b/i.test(def.cost || '')) {
    const n = Math.max(1, amount || 1)
    if (casterKind === 'commander') return { pool: 'commanderAp', amount: n }
    // Officers and Combat Units spend Company AP from the company pool.
    return { pool: 'companyAp', amount: n }
  }

  return { error: 'Ability has no spendable cost.' }
}

File: unity/CommandWarfare/Assets/Scripts/Core/Combat/AbilityCast.cs
using System;
using System.Collections.Generic;
using CommandWarfare.Core.Types;

namespace CommandWarfare.Core.Combat
{
    /// <summary>Port of play/shared/abilityCast.ts — spend resolution for ability casting.</summary>
    public static class AbilityCast
    {
        public enum AbilityPool { None, CommanderAp, CommanderCc, CompanyAp }

        public readonly struct AbilitySpend
        {
            public AbilityPool Pool { get; }
            public int Amount { get; }
            public string Error { get; }

            public bool HasError => Error != null;

            public static AbilitySpend Ok(AbilityPool pool, int amount) =>
                new(pool, amount, null);

            public static AbilitySpend Fail(string error) =>
                new(AbilityPool.None, 0, error);

            AbilitySpend(AbilityPool pool, int amount, string error)
            {
                Pool = pool;
                Amount = amount;
                Error = error;
            }
        }

        public readonly struct AbilityDef
        {
            public string Name { get; }
            public string Type { get; }
            public string Cost { get; }
            public int CostAmount { get; }
            public string CostResource { get; }
            public string UsedBy { get; }

            public AbilityDef(
                string name,
                string type,
                string cost,
                int costAmount,
                string costResource,
                string usedBy)
            {
                Name = name;
                Type = type;
                Cost = cost;
                CostAmount = costAmount;
                CostResource = costResource;
                UsedBy = usedBy;
            }
        }

        static readonly Dictionary<string, string> EffectAliases = new()
        {
            ["Pack Mark"] = "Alpha Mark",
            ["Harden Decree"] = "Harden Order",
            ["Siege Repair"] = "Repair",
            ["Raise Host"] = "Raise Thrall",
            ["Overdrive Pulse"] = "Overdrive",
        };

        public static string ResolveEffectName(string abilityName) =>
            EffectAliases.TryGetValue(abilityName, out var alias) ? alias : abilityName;

        public static bool IsPassive(AbilityDef def)
        {
            var kind = (def.Type ?? "").Trim();
            var cost = (def.Cost ?? "").Trim().ToLowerInvariant();
            return kind == "Passive" || cost == "passive";
        }

        public static bool IsUltimate(AbilityDef def)
        {
            var kind = (def.Type ?? "").Trim();
            var cost = (def.Cost ?? "").Trim().ToLowerInvariant();
            return kind == "Ultimate" || cost == "ultimate";
        }

        public static bool CasterMayUse(AbilityDef def, UnitKind casterKind)
        {
            if (IsPassive(def)) return false;
            var usedBy = (def.UsedBy ?? "").Trim();
            if (casterKind == UnitKind.Commander)
                return IsUltimate(def) || usedBy == "Commander";
            if (casterKind == UnitKind.Officer)
                return !IsUltimate(def) && !CostsCc(def) &&
                       (usedBy == "Officer" || usedBy == "Both");
            return !IsUltimate(def) && !CostsCc(def) &&
                   (string.IsNullOrEmpty(usedBy) || usedBy == "Unit" || usedBy == "Both");
        }

        public static AbilitySpend SpendForCaster(AbilityDef def, UnitKind casterKind)
        {
            if (IsPassive(def)) return AbilitySpend.Fail("Passives are always on.");
            if (IsUltimate(def)) return AbilitySpend.Ok(AbilityPool.None, 0);

            var amount = Math.Max(0, def.CostAmount);
            if (CostsCc(def))
            {
                if (casterKind != UnitKind.Commander)
                    return AbilitySpend.Fail("CC abilities are commander-only.");
                return AbilitySpend.Ok(AbilityPool.CommanderCc, Math.Max(1, amount > 0 ? amount : 1));
            }

            var resource = (def.CostResource ?? "").Trim().ToUpperInvariant();
            var costText = (def.Cost ?? "").ToUpperInvariant();
            if (resource == "AP" || costText.Contains("AP"))
            {
                var n = Math.Max(1, amount > 0 ? amount : 1);
                return casterKind == UnitKind.Commander
                    ? AbilitySpend.Ok(AbilityPool.CommanderAp, n)
                    : AbilitySpend.Ok(AbilityPool.CompanyAp, n);
            }

            return AbilitySpend.Fail("Ability has no spendable cost.");
        }

        static bool CostsCc(AbilityDef def)
        {
            if ((def.CostResource ?? "").Trim().ToUpperInvariant() == "CC") return true;
            var upper = (def.Cost ?? "").ToUpperInvariant();
            if (!upper.Contains("CC") || upper.Contains("COMPANY")) return false;
            return upper.Contains("CC");
        }
    }
}
